test/parallel_api/iterator/permutation_iterator.pass.cpp

```diff
@@ -39,13 +81,490 @@ main()
+    operator()(Policy&& exec, Iterator1 first1, Iterator1 last1, Size n)
+    {
+        for (bool bCallMakeHeap : {false, true})
+        {
+            TestDataTransfer<UDTKind::eKeys, Size> host_keys(*this, n);     // source data for find
+        
+            // Fill full source data set
+            generate_data(host_keys.get(), host_keys.get() + n);
+            if (bCallMakeHeap)
+                ::std::make_heap(host_keys.get(), host_keys.get());
+
+            test_through_permutation_iterator_mibt{first1, n}(
+                [&](auto permItBegin, auto permItEnd)
+                {
+                    const auto result = dpl::is_heap(exec, permItBegin, permItEnd);
+                    // TODO wait_and_throw(exec);
+
+                    const auto expected = std::is_heap(permItBegin, permItEnd);
+
+                    EXPECT_EQ(expected, result, "Wrong result of dpl::is_heap");
+                });
+        }
+    }
+
+    template <typename Policy, typename Iterator1, typename Size>
+    ::std::enable_if_t<!is_random_access_iterator_v<Iterator1>>
+    operator()(Policy&& exec, Iterator1 first1, Iterator1 last1, Size n)
+    {
+    }
+};
+
+template <typename ValueType, template <::std::size_t CallNumber> class invoke_type>
+void
+run_algo_tests()
+{
+    // test_sort : dpl::sort -> __parallel_stable_sort (only for random_access_iterator)
+    test_algo_one_sequence<ValueType, test_sort<ValueType>, invoke_type<0>>();
+
+    // dpl::partial_sort -> __parallel_partial_sort (only for random_access_iterator)
+    test_algo_one_sequence<ValueType, test_partial_sort<ValueType>, invoke_type<1>>();
+
+    // dpl::transform -> __parallel_for (only for random_access_iterator)
+    test_algo_two_sequences<ValueType, test_transform<ValueType>, invoke_type<2>>();
+
+    // dpl::remove_if -> __parallel_transform_scan (only for random_access_iterator)
+    test_algo_one_sequence<ValueType, test_remove_if<ValueType>, invoke_type<3>>();
+
+    // dpl::reduce, dpl::transform_reduce -> __parallel_transform_reduce (only for random_access_iterator)
+    test_algo_one_sequence<ValueType, test_transform_reduce<ValueType>, invoke_type<4>>();
+
+    // dpl::merge, dpl::inplace_merge -> __parallel_merge
+    test_algo_two_sequences<ValueType, test_merge<ValueType>, invoke_type<5>>();
+
+    // dpl::find, dpl::find_if, dpl::find_if_not -> __parallel_find -> _parallel_find_or
+    test_algo_one_sequence<ValueType, test_find<ValueType>, invoke_type<6>>();
+
+    // dpl::is_heap, dpl::includes -> __parallel_or -> _parallel_find_or
+    test_algo_one_sequence<ValueType, test_is_heap<ValueType>, invoke_type<6>>();
+}
+
+template <::std::size_t CallNumber>
+using invoke_on_all_host_policies_wrapper = invoke_on_all_host_policies /*<CallNumber>*/;
+
+template <::std::size_t CallNumber>
+using invoke_on_all_hetero_policies_wrapper = invoke_on_all_hetero_policies<CallNumber>;
+
+#endif // TEST_DPCPP_BACKEND_PRESENT
+
+int
+main()
+{
+#if TEST_DPCPP_BACKEND_PRESENT
```

Comment:
The following code block  confuses me....
`#if TEST_DPCPP_BACKEND_PRESENT` tells that code block is indented just for DPPP backend (hetero backend).
But on the line 564 we call the tests with the host policy - `invoke_on_all_host_policies_wrapper`.
and on the line 567 we dont call the tests with hetero policy (?)

Reply:
Fixed